File: app/src/main/java/com/esorokin/lantector/model/network/api/handler/ObservableCallAdapter.java
package com.esorokin.lantector.model.network.api.handler;

import java.lang.reflect.Type;

import android.support.annotation.NonNull;

import io.reactivex.Observable;
import io.reactivex.Scheduler;
import retrofit2.Call;
import retrofit2.CallAdapter;

class ObservableCallAdapter<BaseResponse> implements CallAdapter<BaseResponse, Observable<BaseResponse>> {
	private final Type responseType;
	private final NetworkErrorAdapter networkErrorAdapter;
	private final HttpResponseAdapter<BaseResponse> httpResponseHandler;
	private final Scheduler scheduler;

	public ObservableCallAdapter(@NonNull Type responseType,
	                             @NonNull NetworkErrorAdapter networkErrorAdapter,
	                             @NonNull HttpResponseAdapter<BaseResponse> httpResponseAdapter,
	                             @NonNull Scheduler scheduler) {
		this.responseType = responseType;
		this.networkErrorAdapter = networkErrorAdapter;
		this.httpResponseHandler = httpResponseAdapter;
		this.scheduler = scheduler;
	}

	@Override
	public Type responseType() {
		return responseType;
	}

	@Override
	public Observable<BaseResponse> adapt(Call<BaseResponse> call) {
		return new CallExecuteObservable<>(call)
				.subscribeOn(scheduler)
				.onErrorResumeNext((Throwable throwable) -> Observable.error(networkErrorAdapter.adaptNetworkError(throwable)))
				.map(httpResponseHandler::adaptHttpResponse);
	}
}

File: app/src/main/java/com/esorokin/lantector/model/network/api/handler/ObservableCallAdapter.kt
package com.esorokin.lantector.model.network.api.handler

import io.reactivex.Observable
import io.reactivex.Scheduler
import retrofit2.Call
import retrofit2.CallAdapter
import java.lang.reflect.Type

internal class ObservableCallAdapter<BaseResponse>(
        private val networkScheduler: Scheduler,
        private val responseType: Type,
        private val networkErrorAdapter: NetworkErrorAdapter,
        private val httpResponseHandler: HttpResponseAdapter<BaseResponse>) : CallAdapter<BaseResponse, Observable<BaseResponse>> {

    override fun responseType(): Type {
        return responseType
    }

    override fun adapt(call: Call<BaseResponse>): Observable<BaseResponse> {
        return CallExecuteObservable(call)
                .subscribeOn(networkScheduler)
                .onErrorResumeNext { throwable: Throwable -> Observable.error(networkErrorAdapter.adaptNetworkError(throwable)) }
                .map(httpResponseHandler::adaptHttpResponse)
    }
}
